R/ltcad.R

```diff
@@ -0,0 +1,133 @@
+#' @method epidist_prepare epidist_ltcad
+#' @family ltcad
+#' @export
+epidist_prepare.epidist_ltcad <- function(data) {
+  id <- obs_t <- obs_at <- NULL
+  ptime_lwr <- pwindow_upr <- stime_lwr <- ptime_upr <- NULL
+  data <- data.table::as.data.table(data)
+  data[, id := seq_len(.N)]
+  data[, obs_t := obs_at - ptime_lwr]
+  data[, pwindow_upr := ifelse(
+    stime_lwr < ptime_upr, ## if overlap
+    stime_upr - ptime_lwr,
+    ptime_upr - ptime_lwr
+  )]
+  data[, woverlap := as.numeric(stime_lwr < ptime_upr)]
+  data[, swindow_upr := stime_upr - stime_lwr]
+  data[, delay_central := stime_lwr - ptime_lwr]
+  data[, row_id := seq_len(.N)]
+  
+  if (nrow(data) > 1) {
+    data <- data[, id := as.factor(id)]
+  }
+  
+  return(data)
+}
+
+#' @method epidist_priors epidist_ltcad
+#' @family ltcad
+#' @export
+epidist_priors.epidist_ltcad <- function(data) {
+  return(NULL)
+}
+
+#' Define a formula for the ltcad model
+#' 
+#' @param delay_central Formula for the delay mean. Defaults to intercept only.
+#' @param sigma Formula for the delay standard deviation. Defaults to intercept
+#' only.
+#' @method epidist_formula epidist_ltcad
+#' @family ltcad
+#' @export
+epidist_formula.epidist_ltcad <- function(data, delay_central = ~ 1,
+                                          sigma = ~ 1) {
+  delay_equation <- paste0(
+    "delay_central | vreal(obs_t, pwindow_upr, swindow_upr)",
+    paste(delay_central, collapse = " ")
+  )
+
+  sigma_equation <- paste0("sigma", paste(sigma, collapse = " "))
+  form <- brms::bf(as.formula(delay_equation), as.formula(sigma_equation))
+  return(form)
+}
+
+#' @method epidist_family epidist_ltcad
+#' @family ltcad
+#' @export
+epidist_family.epidist_ltcad <- function(data, family = "lognormal") {
+  brms::custom_family(
+    paste0("latent_", family),
+    dpars = c("mu", "sigma"),
+    links = c("identity", "log"),
+    lb = c(NA, 0),
+    ub = c(NA, NA),
+    type = "real",
+    vars = c("pwindow", "swindow", "vreal1"),
+    loop = FALSE
+  )
+}
+
+#' @method epidist epidist_ltcad
+#' @family ltcad
+#' @export
+epidist.epidist_ltcad <- function(data, formula = epidist_formula(data),
+                                  family = epidist_family(data),
+                                  priors = epidist_priors(data),
+                                  fn = brms::brm,
+                                  ...) {
+  
+  stanvars_version <- epidist_version_stanvar()
+  
+  stanvars_functions <- brms::stanvar(
```

Comment:
Can you summarise what advantages and disadvantages you see having the stan code loading and overwriting within this function vs in its own function and break those down either based on using the old arg input and the other option (an internal function)

Reply:
I've moved to using its own function. Added issue to discuss this here: https://github.com/epinowcast/epidist/issues/78